batchflow/models/tf/layers/conv.py

```diff
@@ -1,199 +1,211 @@
 """ Contains convolutional layers """
+from functools import partial
 import tensorflow as tf
+import tensorflow.keras.layers as K #pylint: disable=import-error
 
-CONV_LAYERS = {
-    1: tf.layers.conv1d,
-    2: tf.layers.conv2d,
-    3: tf.layers.conv3d
-}
+from .utils import add_as_function
 
-def conv(inputs, *args, **kwargs):
-    """ Nd convolution layer. Just a wrapper around ``tf.layers.conv1d``, ``conv2d``, ``conv3d``.
 
-    Parameters
-    ----------
-    inputs : tf.Tensor
-        input tensor
+
+@add_as_function
+class Conv:
+    """ Nd convolution layer. Just a wrapper around TensorFlow layers for corresponding dimensions.
 
     See also
     --------
-    `tf.layers.conv1d <https://www.tensorflow.org/api_docs/python/tf/layers/conv1d>`_,
-    `tf.layers.conv2d <https://www.tensorflow.org/api_docs/python/tf/layers/conv2d>`_,
-    `tf.layers.conv3d <https://www.tensorflow.org/api_docs/python/tf/layers/conv3d>`_
+    `tf.layers.conv1d <https://www.tensorflow.org/api_docs/python/keras/layers/Conv1D>`_,
+    `tf.layers.conv2d <https://www.tensorflow.org/api_docs/python/keras/layers/Conv2D>`_,
+    `tf.layers.conv3d <https://www.tensorflow.org/api_docs/python/keras/layers/Conv3D>`_
     """
-    dim = inputs.shape.ndims - 2
-    layer_fn = CONV_LAYERS[dim]
-    return layer_fn(inputs, *args, **kwargs)
+    LAYERS = {
+        1: K.Conv1D,
+        2: K.Conv2D,
+        3: K.Conv3D
+    }
+
+    def __init__(self, *args, **kwargs):
+        self.args, self.kwargs = args, kwargs
 
+    def __call__(self, inputs):
+        layer_fn = self.LAYERS[inputs.shape.ndims - 2]
+        return layer_fn(*self.args, **self.kwargs)(inputs)
 
-def conv1d_transpose(inputs, filters, kernel_size, strides=1, padding='valid', data_format='channels_last',
-                     **kwargs):
-    """ Transposed 1D convolution layer
+
+
+@add_as_function
+class Conv1DTranspose:
+    """ Transposed 1D convolution layer.
 
     Parameters
     ----------
-    inputs : tf.Tensor
-        input tensor
     filters : int
-        number of filters in the ouput tensor
+        Number of filters in the ouput tensor.
     kernel_size : int
-        kernel size
+        Kernel size.
     strides : int
-        convolution stride. Default is 1.
+        Convolution stride. Default is 1.
+    """
+    def __init__(self, filters, kernel_size, strides=1, padding='valid', data_format='channels_last', *args, **kwargs):
+        self.filters, self.kernel_size, self.strides = filters, kernel_size, strides
+        self.padding, self.data_format = padding, data_format
+        self.args, self.kwargs = args, kwargs
```

Comment:
I feel very uneasy about converting all these parameters from fields to arguments in ConvBlock and vice versa here. If we need to change parameters list we'll have to check 4 places: ConvBlock init, inside case branch, here and in `__call__ `

What about introducing a small base class
```
class ParameterBuilder:
    parameters = []

    def __init__(self, *args, **kwargs):
        for p in self.parameters:
            self.__setattr__(p, kwargs.pop(p))

        self.args, self.kwargs = args, kwargs
```
we could use it like this:
```
class Conv1DTransposeNn(ParameterBuilder):
    parameters = ['filters', 'output_shape', 'strides', 'padding', 'data_format']

    def my_print(self):
        print(self.padding)
```

and in ConvBlock we could assemble parameters like this:
```
args = {p: getattr(self, p) for p in layer_fn.parameters if hasattr(self, p)}
```

and we will perhaps have to rename `layer_fn` to `layer_cls` in ConvBlock to make it more clear

Reply:
I achieved quite similar (in some ways) functional via `Layer` class, though it is not similar (in some other ways) to the class you described.
You explicitly define list of parameters and then build `__init__` based on this list. This approach lacks defaults for parameters (which can be quite easily added), and it would be hard to add some additional checks in `__init__` (like checking that `conv`-filters are positive ints).
My approach is to define `__init__` explicitly and create list of parameters via `property`. I think that explicit `__init__` helps while looking at class definitions, although your suggestion is great too